src/lib/misc/roughtime/roughtime.cpp

```diff
@@ -0,0 +1,370 @@
+
+std::array<uint8_t, request_min_size> encode_request(const std::array<uint8_t, 64>& nonce)
+  {
+  std::array<uint8_t, request_min_size> buf = {2,0,0,0,64,0,0,0,'N','O','N','C','P','A','D',0xff};
+  std::memcpy(buf.data() + 16, nonce.data(), nonce.size());
+  std::memset(buf.data() + 16 + nonce.size(), 0, buf.size() - 16 - nonce.size());
+  return buf;
+  }
+
+Response Response::from_bits(const std::vector<uint8_t>& response,
+                             const std::array<uint8_t, 64>& nonce)
+  {
+  const auto response_v = split(response);
+  const auto cert = split(get_v(response_v,"CERT"));
+  const auto cert_dele = get<std::array<uint8_t,72>>(cert,"DELE");
+  const auto cert_sig =  get<std::array<uint8_t,64>>(cert,"SIG");
+  const auto cert_dele_v = split(cert_dele);
+  const auto srep = get_v(response_v,"SREP");
+  const auto srep_v = split(srep);
+
+  const auto cert_dele_pubk = get<std::array<uint8_t,32>>(cert_dele_v,"PUBK");
+  const auto sig  = get<std::array<uint8_t,64>>(response_v,"SIG");
+  if(!verify_signature(cert_dele_pubk, srep, sig))
+    throw Invalid_Argument("response signature invalid");
+
+  const auto indx = get<uint32_t>(response_v,"INDX");
+  const auto path = get_v(response_v,"PATH");
+  const auto srep_root = get<std::array<uint8_t,64>>(srep_v, "ROOT");
+  const auto size = path.size();
+  const auto levels = size/64;
+
+  if(size % 64)
+    throw Invalid_Argument("Merkle tree path size must be multiple of 64 bytes");
+  if(indx >= std::exp2(levels))
+    throw Invalid_Argument("Merkle tree path is too short");
+
+  auto hash = hashLeaf(nonce);
+  auto index = indx;
+  auto level = 0u;
+  while(level < levels)
+    {
+    if(index&1)
+      hash = hashNode(*(const std::array<uint8_t, 64>*)(path.data() + level*64), hash);
+    else
+      hash = hashNode(hash, *(const std::array<uint8_t, 64>*)(path.data() + level*64));
+    ++level;
+    index>>=1;
+    }
+
+  if (srep_root != hash)
+    throw Invalid_Argument("nonce verification failed");
+
+  const auto cert_dele_maxt = get<sys_microseconds64>(cert_dele_v,"MAXT");
+  const auto cert_dele_mint = get<sys_microseconds64>(cert_dele_v,"MINT");
+  const auto srep_midp = get<sys_microseconds64>(srep_v,"MIDP");
+  const auto srep_radi = get<microseconds32>(srep_v,"RADI");
+  if(srep_midp < cert_dele_mint)
+    throw Invalid_Argument("midpoint earlier than delegation start");
+  if(srep_midp > cert_dele_maxt)
+    throw Invalid_Argument("midpoint later than delegation end");
+  return {cert_dele, cert_sig, srep_midp, srep_radi};
+  }
+
+bool Response::validate(const Ed25519_PublicKey& pk) const
+  {
+  const char context[] = "RoughTime v1 delegation signature--";
+  PK_Verifier verifier(pk, "Pure");
+  verifier.update((const uint8_t*)context, sizeof(context)); //add context including \0
+  verifier.update(m_cert_dele.data(), m_cert_dele.size());
+  return verifier.check_signature(std::vector<uint8_t>(m_cert_sig.data(), m_cert_sig.data() + m_cert_sig.size()));
```

Comment:
Instead of creating a new vector here just pass the pointer+length as two params: `m_cert_sig.data(), m_cert_sig.size()`